applications/serial_lte_modem/src/slm_at_httpc.c

```diff
@@ -0,0 +1,736 @@
+		return -errno;
+	}
+
+	return 0;
+}
+
+static int server_connect(const char *host, int sec_tag)
+{
+	int fd = -1;
+	int err;
+
+	if (host == NULL) {
+		LOG_ERR("Empty remote host.");
+		return -EINVAL;
+	}
+
+	if ((httpc.sec_transport == true) && (sec_tag == -1)) {
+		LOG_ERR("Empty secure tag.");
+		return -EINVAL;
+	}
+
+	/* Attempt IPv6 connection if configured, fallback to IPv4 */
+	fd = resolve_and_connect(AF_INET6, host, sec_tag);
+	if (fd < 0) {
+		fd = resolve_and_connect(AF_INET, host, sec_tag);
+	}
+
+	if (fd < 0) {
+		LOG_ERR("Fail to resolve and connect");
+		return -EINVAL;
+	}
+	LOG_INF("Connected to %s", log_strdup(host));
+
+	/* Set socket timeout, if configured */
+	err = socket_timeout_set(fd);
+	if (err) {
+		return err;
+	}
+
+	return fd;
+}
+
+static void response_cb(struct http_response *rsp,
+			enum http_final_call final_data,
+			void *user_data)
+{
+	static size_t data_received;
+
+	data_received += rsp->data_len;
+	if (final_data == HTTP_DATA_MORE) {
+		LOG_DBG("Partial data received (%zd bytes)", rsp->data_len);
+		if (data_received == HTTPC_BUF_LEN) {
+			sprintf(rsp_buf, "#XHTTPCRSP:%d,1\r\n",
+				HTTPC_BUF_LEN);
+			rsp_send(rsp_buf, strlen(rsp_buf));
+			rsp_send(data_buf, HTTPC_BUF_LEN);
+			data_received = 0;
+		}
+	} else if (final_data == HTTP_DATA_FINAL) {
+		LOG_DBG("All the data received (%zd bytes)", data_received);
+		sprintf(rsp_buf, "#XHTTPCRSP:%d,0\r\n", data_received);
+		rsp_send(rsp_buf, strlen(rsp_buf));
+		rsp_send(data_buf, data_received);
+		httpc.rsp_completed = true;
+		data_received = 0;
+	}
+}
+
+static int headers_cb(int sock, struct http_request *req, void *user_data)
+{
+	size_t len;
+	int ret = 0;
+
+	len = strlen(httpc.headers);
+	if (len > 0) {
+		ret = send(sock, httpc.headers, len, 0);
+		if (ret < 0) {
+			LOG_ERR("send header fail: %d", ret);
+			return ret;
+		}
+		LOG_DBG("send header: %d", ret);
+	}
+
+	return ret;
+}
+
+static int payload_cb(int sock, struct http_request *req, void *user_data)
+{
+	size_t total_sent = 0;
+
+	if (httpc.pl_len > 0) {
+		sprintf(rsp_buf, "#XHTTPCREQ:1\r\n");
+		rsp_send(rsp_buf, strlen(rsp_buf));
+		do {
+			/* Wait until payload is ready */
+			LOG_DBG("wait until payload is ready");
+			k_sem_take(&http_req_sem, K_FOREVER);
+			if (httpc.pl_len == 0) {
+				LOG_INF("abort sending payload");
+				httpc.pl_sent = 0;
+				return -ECONNABORTED;
+			}
+			if (total_sent + httpc.pl_to_send > httpc.pl_len) {
+				LOG_WRN("send unexpected payload");
+				httpc.pl_to_send = httpc.pl_len - total_sent;
+			}
+			while (httpc.pl_sent < httpc.pl_to_send) {
+				ssize_t ret;
+
+				ret = send(sock, data_buf + httpc.pl_sent,
+					   MIN(httpc.pl_to_send - httpc.pl_sent,
+					   NET_IPV4_MTU), 0);
+				if (ret < 0) {
+					LOG_ERR("send fail: %d", ret);
+					httpc.pl_len = 0;
+					httpc.pl_sent = 0;
+					k_sem_give(&http_data_sem);
+					return -errno;
+				}
+				LOG_DBG("send %d bytes payload", total_sent);
```

Comment:
Is this in a right place? `total_sent` was not increased yet. Or perhaps it should only print out the amount of bytes sent in a single iteration (`ret`).

Reply:
The intention is to print the amount of single iteration. Fixed by replace total_sent as ret.